csharp/src/Ice/IncomingRequestFrame.cs

```diff
@@ -0,0 +1,45 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+using System;
+using System.Collections.Generic;
+
+namespace Ice
+{
+    using Context = Dictionary<string, string>;
+
+    /// <summary>Represents a request protocol frame received by the application.</summary>
+    // TODO: IncomingRequestFrame should derive from InputStream
+    public sealed class IncomingRequestFrame
+    {
+        public InputStream InputStream { get; }
+
+        // TODO: combine Current and RequestFrameHeader in some way?
+        public Current Current { get; }
+
+        /// <summary>The payload of this request frame. The bytes inside the payload should not be written to;
+        /// they are writable because of the <see cref="System.Net.Sockets.Socket"/> methods for sending.</summary>
+        // TODO: describe how long this payload remains valid once we add memory pooling.
+        public ArraySegment<byte> Payload { get; }
+
+        /// <summary>Take the payload from this response frame. After calling this method, the payload can no longer
+        /// be read.</summary>
+        public ArraySegment<byte> TakePayload()
+        {
+            // TODO: make this method destructive with memory pooling.
+            return Payload;
+        }
+
+        public IncomingRequestFrame(InputStream inputStream, Current current)
+        {
+            InputStream = inputStream;
+            Current = current;
+
+            // TODO: works only when Payload called first before reading anything. Need a better version!
```

Comment:
Is this because it depends on the input stream position?

Reply:
Yes.

Reply:
It is somehow related to issue with GetEnumerator

https://github.com/zeroc-ice/ice/blob/master/csharp/src/Ice/InputStream.cs#L2121

My idea was to have a inputstream view with a position that doesn't depend on the underlying stream.

Reply:
I think the solution is for the InputStream to "know" where this payload is and be capable of returning it upon request. See Payload and TakePayload().